niworkflows/utils/spaces.py

```diff
@@ -0,0 +1,418 @@
+
+    >>> sp.insert(0, ('MNIPediatricAsym', {'cohort': '3'}))
+    >>> sp.get_std_spaces(dim=(3,))
+    ['MNIPediatricAsym:cohort-3',
+     'MNI152NLin2009cAsym',
+     'MNIPediatricAsym:cohort-2',
+     'MNIPediatricAsym:cohort-1']
+
+    >>> sp.insert(0, ('MNIPediatricAsym', {'cohort': '3'}))
+    Traceback (most recent call last):
+      ...
+    ValueError: space ...
+
+    """
+
+    __slots__ = ('_spaces')
+    standard_spaces = tuple(_tfapi.templates())
+    """List of supported standard reference spaces."""
+
+    @staticmethod
+    def check_space(space):
+        """Build a :class:`Space` object."""
+        spec = {}
+        if not isinstance(space, str):
+            try:
+                spec = space[1] or {}
+            except IndexError:
+                pass
+            except TypeError:
+                space = (None, )
+
+            space = space[0]
+        return Space(space, spec)
+
+    def __init__(self, spaces=None):
+        """
+        Maintain the bookkeeping of spaces and templates.
+
+        Internal spaces are normalizations required for pipeline execution which
+        can vary based on user arguments.
+        Output spaces are desired user outputs.
+        """
+        self._spaces = []
+        if spaces is not None:
+            if isinstance(spaces, str):
+                spaces = [spaces]
+            self.__add__(spaces)
+
+    def __add__(self, b):
+        """Append a list of transforms to the internal list."""
+        if not isinstance(b, (list, tuple)):
+            raise TypeError('Must be a list.')
+
+        for space in b:
+            self.append(space)
+        return self
+
+    def __contains__(self, item):
+        """Implement the ``in`` builtin."""
+        if not self._spaces:
+            return False
+        item = self.check_space(item)
+        return any([i == item for i in self._spaces])
+
+    def __repr__(self):
+        """Representation of this object."""
+        return 'Imaging spaces: %s' % ', '.join(self._spaces)
+
+    @property
+    def spaces(self):
+        """Get all specified spaces."""
+        return self._spaces
+
+    @spaces.setter
+    def spaces(self, value):
+        self._spaces = value
+
+    def add(self, value):
+        """Add one more space, without erroring if it exists."""
+        if value not in self:
+            self.spaces += [self.check_space(value)]
+
+    def append(self, value):
+        """Concatenate one more space."""
+        if value not in self:
+            self.spaces += [self.check_space(value)]
+            return
+
+        raise ValueError('space "%s" already in spaces.' % str(value))
+
+    def insert(self, index, value, error=True):
+        """Concatenate one more space."""
+        if value not in self:
+            self.spaces.insert(index, self.check_space(value))
+        elif error is True:
+            raise ValueError('space "%s" already in spaces.' % str(value))
+
+    def get_std_spaces(self, dim=(2, 3)):
```

Comment:
having the `only_names` keyword like `get_nonstd_spaces` would be useful

Reply:
this method intendedly provides only names. It could have some use below, in ``get_templates`` (although missing out the spec kind of defeats the purpose). Actually, I would think more of removing the ``only_names`` from ``get_nonstd_spaces``.

Reply:
I'm a still a bit confused how this plans on separating user requested spaces from internally added ones.

the way most outputnodes are currently set up, a list of all spaces are returned. how can we then filter out only the requested spaces with this class?

Reply:
Let's break it down into smaller pieces.

- Volumes (spatial normalization):
  In this case, we just need to feed this iterable with the right information https://github.com/poldracklab/smriprep/blob/master/smriprep/workflows/norm.py#L141.
  I might be blindsided at this moment, but I believe having template ID + cohort should suffice. Since it is an iterable, Nipype will take care of all the synchronization (and that is how it actually works currently).

- Surfaces:
  I will need to revise things, but my impression is that sMRIPrep does not need anything about surfaces, other than whether recon-all is run or not.
  In fMRIPrep proper, we need surface output spaces to decide whether we generate derivatives or not (therefore, no concept of "internal" space is necessary). With this way of thinking, we will be able to run recon-all but abstain from generating derivatives if the user does not explicitly request them.

As I said, I'll need to go through https://github.com/poldracklab/smriprep/pull/159 to make sure this all makes sense and works out.

Reply:
> I might be blindsided at this moment, but I believe having template ID + cohort should suffice. Since it is an iterable, Nipype will take care of all the synchronization (and that is how it actually works currently).

We'll have to pass in specs as well, to fully differentiate the spaces. IIRC, duplicate iterables will overwrite working directories.

Reply:
https://github.com/mgxd/smriprep/pull/1/files#diff-816eb5ff17b071ecf371cf48577d5a3cR246

``Space.fullname`` is doing the magic - unique template ID (+cohort) are passed.